### PR TITLE
OKTA-643929 - secret folders

### PR DESCRIPTION
**Pull request by SajanAlexander-okta**
This implements the secret folders resource / datasource

### CODE REVIEW

File: oktapam/constants/descriptions/links.go [hunk 1]
```diff
@@ -15,10 +15,10 @@ const (
...
-	LinkResourceGroupServerEnrollmentToken = "For details, see [Server Enrollment](https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/server-agent/pam-enroll-a-server.htm)."
-	LinkSecurityPolicy                     = "For details, see [Manage security policy](https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/pam-policy.htm)."
+	LinkPasswordSettings                   = "For details, see [Projects](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-projects.htm)."
+	LinkResourceGroup                      = "For details, see [Resource Groups](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-resource-groups.htm)."
+	LinkResourceGroupProject               = "For details, see [Projects](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-projects.htm)."
+	LinkResourceGroupServerEnrollmentToken = "For details, see [Server Enrollment](https://help.okta.com/oie/en-us/content/topics/privileged-access/server-agent/pam-enroll-a-server.htm)."
```
Comment (isaacdurham-okta) [inline oktapam/constants/descriptions/links.go:21]:
```suggestion
	LinkResourceGroupServerEnrollmentToken = "For details, see [Server Enrollment](https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-server-enrollment)."
```

File: oktapam/constants/descriptions/links.go [hunk 2]
```diff
@@ -15,10 +15,10 @@ const (
...
-	LinkResourceGroup                      = "For details, see [Resource Groups](https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/pam-resource-groups.htm)."
-	LinkResourceGroupProject               = "For details, see [Projects](https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/pam-projects.htm)."
-	LinkResourceGroupServerEnrollmentToken = "For details, see [Server Enrollment](https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/server-agent/pam-enroll-a-server.htm)."
-	LinkSecurityPolicy                     = "For details, see [Manage security policy](https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/pam-policy.htm)."
+	LinkPasswordSettings                   = "For details, see [Projects](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-projects.htm)."
+	LinkResourceGroup                      = "For details, see [Resource Groups](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-resource-groups.htm)."
```
Comment (isaacdurham-okta) [inline oktapam/constants/descriptions/links.go:19]:
```suggestion
	LinkResourceGroup                      = "For details, see [Resource Groups](https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-resource-groups)."
```

File: oktapam/constants/descriptions/links.go [hunk 3]
```diff
@@ -15,10 +15,10 @@ const (
...
+	LinkResourceGroupProject               = "For details, see [Projects](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-projects.htm)."
+	LinkResourceGroupServerEnrollmentToken = "For details, see [Server Enrollment](https://help.okta.com/oie/en-us/content/topics/privileged-access/server-agent/pam-enroll-a-server.htm)."
+	LinkSecretFolder                       = "For details, see [Secret Folders](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-secrets-folder.htm)"
+	LinkSecurityPolicy                     = "For details, see [Manage security policy](https://help.okta.com/oie/en-us/content/topics/privileged-access/pam-policy.htm)."
```
Comment (isaacdurham-okta) [inline oktapam/constants/descriptions/links.go:23]:
```suggestion
	LinkSecurityPolicy                     = "For details, see [Security policy](https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-policy)."
```

File: oktapam/constants/descriptions/resources.go [hunk 4]
```diff
@@ -15,22 +15,23 @@ var (
 	ResourceADServerSyncTaskSettings = fmt.Sprintf("Settings for an Active Directory (AD) task that discovers servers. %s", LinkADServerDiscovery)
 	ResourceADUserSyncTaskSettings   = fmt.Sprintf("Settings for an Active Directory (AD) task that discovers user SIDs. %s", LinkADUserDiscovery)
 	ResourceGatewaySetupToken        = fmt.Sprintf("A token used to enroll a Gateway. %s", LinkGatewaySetupToken)
-	ResourceGroup                    = fmt.Sprintf("A set of ASA Users. %s", LinkGroup)
+	ResourceGroup                    = fmt.Sprintf("A set of ASA/PAM Users. %s", LinkGroup)
```
[the remark this answers is not in the document]
Comment (SajanAlexander-okta) [reply oktapam/constants/descriptions/resources.go:18]: yep - so it's not a PAM Resource Group, it's the PAM/ASA Group Terraform Resource

File: oktapam/constants/descriptions/prefixes.go [hunk 5]
```diff
@@ -3,5 +3,6 @@ package descriptions
 const (
 	PrefixBeta = "Beta Feature:"
 	PrefixLEA  = "Limited Early Access Feature:"
-	PrefixDev  = "Feature under development:"
+	PrefixDev  = "Under Development Feature :"
```
[the remark this answers is not in the document]
Comment (isaacdurham-okta) [reply oktapam/constants/descriptions/prefixes.go]: Does this need the space before the colon?

**Review by isaacdurham-okta — APPROVED** [review]

File: docs/data-sources/password_settings.md [hunk 6]
```diff
@@ -3,12 +3,12 @@
...
 subcategory: ""
 description: |-
-  Returns an existing Password Policy for a PAM Project.  For details, see Projects https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/pam-projects.htm
+  Returns an existing Password Policy for a PAM Project.  For details, see Projects https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-projects.
 ---
 
 # oktapam_password_settings (Data Source)
 
-Returns an existing Password Policy for a PAM Project.  For details, see [Projects](https://help.okta.com/en/programs/opa-pam/Content/Topics/privileged-access/pam-projects.htm)
+Returns an existing Password Policy for a PAM Project.  For details, see [Projects](https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-projects).
```
Comment (mariovillaplana-okta) [inline docs/data-sources/password_settings.md:11]: This link takes me to the generic OIE docs: https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-projects

File: docs/resources/secret_folder.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "oktapam_secret_folder Resource - terraform-provider-oktapam"
+subcategory: ""
+description: |-
+  A file-system like construct that contains secrets or nested secret folders. For details, see Secret Folders https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-secrets-folder
+---
+
+# oktapam_secret_folder (Resource)
+
+A file-system like construct that contains secrets or nested secret folders. For details, see [Secret Folders](https://help.okta.com/okta_help.htm?type=oie&id=ext-pam-secrets-folder)
```
Comment (mariovillaplana-okta) [inline docs/resources/secret_folder.md:11]: This link also just takes me to generic OIE docs

Comment (SajanAlexander-okta) [reply docs/resources/secret_folder.md:11]: @isaacdurham-okta informed me that the doc links will be live next week, but we should be using this format

File: docs/data-sources/password_settings.md (hunk 6 above)
Comment (isaacdurham-okta) [reply docs/data-sources/password_settings.md:11]: These alias links won't work until we republish the docs again next week.

File: oktapam/client/secret_folder.go [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+			return nil, callErr
+		}
+
+		for _, s := range resp.List {
+			if s.Type == pam.SecretType_FOLDER {
+				list = append(list, ConvertSecretOrFolderListResponseToSecretFolder(s))
+			}
+		}
+
+		linkHeader := httpResp.Header.Get("Link")
```
[the remark this answers is not in the document]
Comment (SajanAlexander-okta) [reply oktapam/client/secret_folder.go:174]: there's enough of a difference (e.g. valid http codes, filtering on the data returned, etc) in the few things that i have done already that i'd want to defer that work until we a handle on what else might be different

File: oktapam/client/user.go [hunk 9]
```diff
@@ -105,6 +106,65 @@ type UsersListResponse struct {
 	Users []User `json:"list"`
 }
 
+func ListServiceUsers(ctx context.Context, sdkClient SDKClientWrapper) ([]User, error) {
```
[the remark this answers is not in the document]
Comment (SajanAlexander-okta) [reply oktapam/client/user.go:109]: Yes, not sure if the agent/client supports them fully, but the platform supports using them.  The provider relies on them.

File: oktapam/constants/descriptions/resources.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (SajanAlexander-okta) [reply oktapam/constants/descriptions/resources.go:18]: I'm not opposed to that, but i'd rather push it into a different PR.

**Review by mariovillaplana-okta — APPROVED** [review]

File: oktapam/client/secret_folder.go [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+
+	return m
+}
+
+func ListTopLevelSecretFolders(ctx context.Context, sdkClient SDKClientWrapper) ([]SecretFolder, error) {
+	var list []SecretFolder
+	request := sdkClient.SDKClient.SecretsAPI.ListTopLevelSecretFoldersForUser(ctx, sdkClient.Team)
+
+	for {
+		resp, httpResp, callErr := request.Execute()
```
Comment (sachinsaxena-okta) [inline oktapam/client/secret_folder.go:54]: I think we can take out the logic to handle link headers in a common method. We can convert API response outside that common method. If you want, we do it in separate PR too.

File: oktapam/client/secret_folder.go [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+	"github.com/okta/terraform-provider-oktapam/oktapam/logging"
+	"github.com/tomnomnom/linkheader"
+)
+
+const secretFolderSeparator = "/"
+
+// SecretFolder represents a secret folder entity.  We use this instead of the PAM SDK response since
+// there are multiple API response types which are used for secret folders and this provides a common
+// structure that can be used
+type SecretFolder struct {
```
Comment (sachinsaxena-okta) [inline oktapam/client/secret_folder.go:21]: I didn't look at the spec, but can't we use same schema in the spec too ?

File: oktapam/client/secret_folder.go [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+	resolveSecretOrFolderRequest := pam.NewResolveSecretOrFolderRequest(
+		pam.SecretResolveParent{Id: &resourceGroupID, Type: pam.NamedObjectType_RESOURCE_GROUP.Ptr()},
+		pam.SecretResolveParent{Id: &projectID, Type: pam.NamedObjectType_PROJECT.Ptr()},
+	)
+
+	resolveSecretOrFolderRequest.SetParentFolderPath(parentFolder)
+	resolveSecretOrFolderRequest.SetSecretFolderName(name)
+
+	resp, httpResp, callErr := sdkClient.SDKClient.SecretsAPI.ResolveSecretOrFolder(ctx, sdkClient.Team).ResolveSecretOrFolderRequest(*resolveSecretOrFolderRequest).Execute()
+	if httpResp != nil {
```
Comment (sachinsaxena-okta) [inline oktapam/client/secret_folder.go:115]: May be this can be put as a common method too to handle httpResponse.

File: oktapam/client/secret_folder.go [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+		Name:            &resp.Name,
+		ID:              &resp.Id,
+		Description:     resp.Description.Get(),
+		ResourceGroupID: resp.ResourceGroup.Id,
+		ProjectID:       resp.Project.Id,
+		ParentFolderID:  parentSecretFolderID,
+	}
+}
+
+func (c OktaPAMClient) GetSecretFolder(ctx context.Context, resourceGroupID string, projectID string, id string) (*SecretFolder, error) {
```
Comment (sachinsaxena-okta) [inline oktapam/client/secret_folder.go:235]: I think this is something we are going to change later to use SDK client.

File: oktapam/client/pam_sdk_client.go [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+	"github.com/atko-pam/pam-sdk-go/client/pam"
+)
+
+type SDKClientWrapper struct {
+	SDKClient *pam.APIClient
+	Team      string
+}
+
+func checkStatusCodeFromSDK(resp *http.Response, allowed ...int) (int, error) {
```
Comment (sachinsaxena-okta) [inline oktapam/client/pam_sdk_client.go:16]: I will add later in SDK to return httpStatusCode as part of the API error returned from sdk itself so that we don't need to check it from the raw http response.

File: oktapam/client/secret_folder.go (hunk 11 above)
Comment (SajanAlexander-okta) [reply oktapam/client/secret_folder.go:21]: I think there are some slight differences between the responses.  It's possible that we can merge this into a single spec, but I think that's outside the scope of this PR.

File: oktapam/client/secret_folder.go (hunk 13 above)
Comment (SajanAlexander-okta) [reply oktapam/client/secret_folder.go:235]: yeah, i will do this in a follow-up PR

**Review by sachinsaxena-okta — APPROVED** [review]
